src/main/java/edu/ucla/library/libcal/services/OAuthTokenServiceImpl.java

```diff
@@ -33,42 +35,44 @@ public class OAuthTokenServiceImpl implements OAuthTokenService {
      */
     private final Vertx myVertx;
 
-    /**
-     * The ID of the periodic timer for checking if the OAuth token needs refreshing.
-     */
-    private final long myTimerId;
-
     /**
      * The OAuth authentication provider.
      */
     private final OAuth2Auth myAuthProvider;
 
     /**
-     * LibCal credentials represented as an OAuth token.
+     * The ID of the periodic timer for checking if the OAuth token needs refreshing.
      */
-    private User myToken;
+    private long myTimerId;
 
     /**
      * Creates an instance of the service. Use {@link OAuthTokenService#create(Vertx, JsonObject)} to invoke.
+     * <p>
+     * Note that the return value should not be used directly. Instead use the result of {@code aPromise}.
      *
      * @param aVertx A Vert.x instance
      * @param aConfig A configuration
-     * @param anAuthProvider An OAuth authentication provider
-     * @param aToken A token that was returned by authenticating against {@code anAuthProvider}
+     * @param aPromise A Promise that completes with the service instance if initialization was successful, or fails
+     *        otherwise
      */
-    public OAuthTokenServiceImpl(final Vertx aVertx, final JsonObject aConfig, final OAuth2Auth anAuthProvider,
-            final User aToken) {
+    public OAuthTokenServiceImpl(final Vertx aVertx, final JsonObject aConfig,
+            final Promise<OAuthTokenService> aPromise) {
+        final OAuth2Options options = new OAuth2Options().setFlow(OAuth2FlowType.CLIENT)
+                .setClientId(aConfig.getString(Config.OAUTH_CLIENT_ID))
+                .setClientSecret(aConfig.getString(Config.OAUTH_CLIENT_SECRET))
+                .setSite(aConfig.getString(Config.OAUTH_TOKEN_URL));
+
         myVertx = aVertx;
-        myAuthProvider = anAuthProvider;
-        myToken = aToken;
-        myTimerId = aVertx.setPeriodic(1000, timerId -> {
-            if (myToken.expired()) {
-                refreshTokenIfExpired().onFailure(details -> LOGGER.error(details.getMessage()));
-            }
-        });
+        myAuthProvider = OAuth2Auth.create(aVertx, options);
+
+        authenticateWithRetry(Optional.of(1), 5).compose(token -> {
+            myTimerId = keepTokenFresh(token, 300);
```

Comment:
Instead of naive polling every 1000 seconds to see if the token has expired (and then attempting a refresh), this starts the refresh attempts 5 minutes before the token is set to expire.